### PR TITLE
Remove TODO rename old file to keypair.old(?)

### PR DESCRIPTION
**Pull request by hskang9**
**Summary of changes**
Changes introduced in this pull request:
- I add a utility function `count_files` to count files in a directory
- ? willl be the number of old keystore files in the libp2p directory
- when the key does not work, the program will label the old keystore file and make new one

### CODE REVIEW

File: node/utils/src/lib.rs [hunk 1]
```diff
@@ -66,3 +66,29 @@ where
...
+/// let file_count = count_files(path_to_keystore[0..path_to_keystore.len()-8].to_string());
+/// assert_eq!(file_count, 1);
+/// ```
+pub fn count_files(dir: String) -> Result<u32> {
+    let mut file_count = 0;
+    for _i in read_dir(dir).expect("read files in the directory") {
+        file_count += 1;
+    }
+    Ok(file_count)
+}
```
Comment (austinabell) [inline node/utils/src/lib.rs]: The return type is a result, but you're not passing up the error and instead panicing. The dir parameter does not need to be a String (heap allocated) and can instead be a `&str`. No need for for loop when only taking count.

Consider possibly something like: 
```
pub fn count_files(dir: &str) -> Result<usize> {
    Ok(read_dir(dir)?.enumerate().count())
}
```

File: node/utils/src/lib.rs [hunk 2]
```diff
@@ -66,3 +66,29 @@ where
...
+/// - The process lacks permissions to view the contents.
+/// - The path points at a non-directory file.
+///
+/// # Example
+/// ```
+/// use utils::count_files;
+/// use utils::get_home_dir;
+///
+/// let path_to_keystore = get_home_dir() + "/.forest/libp2p/keypair";
+/// let file_count = count_files(path_to_keystore[0..path_to_keystore.len()-8].to_string());
```
Comment (austinabell) [inline node/utils/src/lib.rs]: why are you indexing to length - 8?

File: node/utils/src/lib.rs [hunk 3]
```diff
@@ -66,3 +66,29 @@ where
...
+/// assert_eq!(file_count, 1);
```
Comment (austinabell) [inline node/utils/src/lib.rs]: file_count here is a result, should unwrap in example in previous line or this

File: node/utils/src/lib.rs [hunk 4]
```diff
@@ -66,3 +66,29 @@ where
...
+/// Will return the number of files in the directory
+///
+/// # Panics
+/// It will panic if:
+/// - The provided path doesn't exist.
+/// - The process lacks permissions to view the contents.
+/// - The path points at a non-directory file.
+///
+/// # Example
+/// ```
```
Comment (austinabell) [inline node/utils/src/lib.rs]:
```suggestion
/// ```no_run
```

Shouldn't have a test that uses non-controlled local file system

File: node/utils/src/lib.rs [hunk 5]
```diff
@@ -66,3 +66,29 @@ where
...
 }
+
+/// Get file count in a certain directory
+/// Will return the number of files in the directory
+///
+/// # Panics
+/// It will panic if:
+/// - The provided path doesn't exist.
+/// - The process lacks permissions to view the contents.
+/// - The path points at a non-directory file.
```
Comment (austinabell) [inline node/utils/src/lib.rs]: Remove this when panic removed in function

File: node/forest_libp2p/src/service.rs [hunk 6]
```diff
@@ -188,6 +189,22 @@ fn get_keypair(log: &Logger) -> Keypair {
                 Err(e) => {
                     info!(log, "Could not decode networking keystore!");
                     trace!(log, "Error {:?}", e);
+                    // Get current number of old files
+                    let file_count =
+                        count_files(path_to_keystore[0..path_to_keystore.len() - 8].to_string());
```
Comment (austinabell) [inline node/forest_libp2p/src/service.rs]: why the -8 indexing here also?

File: node/forest_libp2p/src/service.rs [hunk 7]
```diff
@@ -188,6 +189,22 @@ fn get_keypair(log: &Logger) -> Keypair {
...
+                    match file_count {
+                        Err(e) => {
+                            info!(log, "Error {:?}", &e);
+                        }
+                        Ok(v) => {
+                            fs::rename(
+                                path_to_keystore.clone(),
+                                path_to_keystore.clone() + &format!(".old({:})", v),
+                            )
+                            .expect("label the file as old key");
```
Comment (austinabell) [inline node/forest_libp2p/src/service.rs]: don't panic if this fails, should be recoverable

File: node/utils/tests/files.rs [hunk 8]
```diff
@@ -116,3 +116,23 @@ fn read_from_toml() {
...
 }
+
+#[test]
+fn count_files_in_a_dir() {
+    let msg = "Hello World!";
+    let path = "./test_string_read_file/";
+    let file_name = "out.keystore";
+    if let Err(e) = write_to_file(&msg.as_bytes().to_vec(), &path, file_name) {
+        panic!(e);
+    }
```
Comment (austinabell) [inline node/utils/tests/files.rs]: Can you change this to:

```
    write_to_file(msg.as_bytes(), &path, file_name).unwrap();
```

Sorry that it's set up wrong above in this file, I can make a pass through after to fix, but there is no need to heap allocate the message or manually handle the error in this test

File: node/utils/src/lib.rs (hunk 2 above)
Comment (hskang9) [reply node/utils/src/lib.rs]: Removing the filename for now. I can replace it like "/keypair".len().

File: node/utils/src/lib.rs (hunk 3 above)
Comment (hskang9) [reply node/utils/src/lib.rs]: Nice catch. I forgot to update documentation

File: node/forest_libp2p/src/service.rs (hunk 6 above)
Comment (hskang9) [reply node/forest_libp2p/src/service.rs]: ok

File: node/utils/src/lib.rs (hunk 5 above)
Comment (hskang9) [reply node/utils/src/lib.rs]: ok

File: node/forest_libp2p/src/service.rs (hunk 7 above)
Comment (hskang9) [reply node/forest_libp2p/src/service.rs]: ok

File: node/utils/tests/files.rs (hunk 8 above)
Comment (hskang9) [reply node/utils/tests/files.rs]: ok

File: node/utils/src/lib.rs (hunk 1 above)
Comment (hskang9) [reply node/utils/src/lib.rs]: ok

**Comment by austinabell** [comment]
Hi @hskang9 sorry to do this but I had to refactor some files referenced in this PR for the work we are doing now. Those files are unlikely to change again in the near future if you do intend on finishing this PR. 

Not a problem if not, but to finish you should run `make lint` to add the license headers to missing files and to fix those clippy errors and run `cargo test` to clear up those test failures.

**Comment by hskang9** [comment]
It is okay. It is good to see that the repository is active. I will make changes today.

**Comment by austinabell** [comment]
going to close this for now because I'm assuming you want to abandon this, if you do intend to finish this feel free to open up again :)

**Pull request closed without merging**